File: unity/Assets/Mg01/Core/DamageCalc.cs
// 伤害计算：与 engine/src/DamageCalc.ts 1:1 对应
// War3 公式: damageAfterArmor = damage * mult * (1 - armor*0.06 / (1 + armor*0.06))
// armor < 0 时: 2 - 0.94^(-armor)

using System;

namespace Mg01.Core
{
    public static class DamageCalc
    {
        /// <summary>返回最终扣血值</summary>
        public static float Calc(float rawDamage, DamageType atk, float armor, ArmorType armorType)
        {
            float mult = DamageMatrix.Multiplier(atk, armorType);
            float typed = rawDamage * mult;
            return typed * ArmorReduction(armor);
        }

        /// <summary>War3 护甲减免系数 (0..1+)</summary>
        public static float ArmorReduction(float armor)
        {
            if (armor >= 0f)
            {
                float k = armor * 0.06f;
                return 1f - k / (1f + k);
            }
            return 2f - (float)Math.Pow(0.94, -armor);
        }
    }
}

File: engine/src/DamageCalc.ts
// 伤害计算：查表 + 护甲减免
// War3 公式: damageAfterArmor = damage * multiplier * (1 - armor*0.06 / (1 + armor*0.06))
// 护甲 >= 0 时用上式；< 0 时用 2 - 0.94^(-armor)

import type { DamageType, ArmorType } from "./Enums.ts";
import { damageMatrix } from "./DataLoader.ts";

export class DamageCalc {
  /** 返回最终造成的血量减少值 */
  static calc(
    rawDamage: number,
    atkType: DamageType,
    armor: number,
    armorType: ArmorType
  ): number {
    const multiplier = damageMatrix[atkType]?.[armorType] ?? 1.0;
    const typed = rawDamage * multiplier;
    return typed * DamageCalc.armorReduction(armor);
  }

  /** War3 护甲减免系数 (0~1)  */
  static armorReduction(armor: number): number {
    if (armor >= 0) {
      const k = armor * 0.06;
      return 1 - k / (1 + k);
    }
    return 2 - Math.pow(0.94, -armor);
  }

  /** 只查矩阵倍率，不计护甲 */
  static multiplier(atkType: DamageType, armorType: ArmorType): number {
    return damageMatrix[atkType]?.[armorType] ?? 1.0;
  }
}
